File: python/d02.py
import re
import sys

with open(f"../{"test-" if "-t" in sys.argv else ""}data/d02.txt", "r") as file:
    ranges = [
        range(int(p.split("-")[0]), int(p.split("-")[1]) + 1)
        for p in file.read().strip().split(",")
    ]

invalid_p1 = []
invalid_p2 = []
pattern_p1 = re.compile(r"^([0-9]+)\1$")
pattern_p2 = re.compile(r"^([0-9]+)\1{1,}$")
for r in ranges:
    for i in r:
        if pattern_p1.match(str(i)):
            invalid_p1.append(i)
        if pattern_p2.match(str(i)):
            invalid_p2.append(i)

print("p1", sum(invalid_p1))
print("p2", sum(invalid_p2))

File: rust/src/bin/d02.rs
use anyhow::Result;

fn main() -> Result<()> {
    let input = if cfg!(feature = "test") {
        include_str!("../../../test-data/d02.txt")
    } else {
        include_str!("../../../data/d02.txt")
    };

    println!("part one: {}", part_one(input)?);
    println!("part two: {}", part_two(input)?);
    Ok(())
}

fn parse_input(input: &str) -> Result<Vec<(usize, usize)>> {
    input
        .split(',')
        .map(|s| {
            let (start, end) = s.split_once('-').ok_or(anyhow::anyhow!("invalid input"))?;
            Ok((start.parse()?, end.parse()?))
        })
        .collect()
}

fn solver(input: &str, pattern: regress::Regex) -> Result<usize> {
    let parsed = parse_input(input)?;

    let mut result = 0;
    for (start, end) in parsed {
        result += (start..(end + 1))
            .filter(|i| pattern.find(i.to_string().as_str()).is_some())
            .sum::<usize>();
    }

    Ok(result)
}

fn part_one(input: &str) -> Result<usize> {
    solver(input, regress::Regex::new(r"^([0-9]+)\1$")?)
}

fn part_two(input: &str) -> Result<usize> {
    solver(input, regress::Regex::new(r"^([0-9]+)\1{1,}$")?)
}
